File: stdlib/lua/hull/web/tests/test_pagination.lua
-- test_pagination.lua — Tests for hull.web.pagination

local pagination = require('hull.web.pagination')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then pass = pass + 1
    else fail = fail + 1; print("FAIL: " .. name .. ": " .. tostring(err)) end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- ── from_query ──────────────────────────────────────────────────────

test("from_query defaults to page=1 per_page=20 when no query", function()
    local p = pagination.from_query({})
    assert_eq(p.page, 1)
    assert_eq(p.per_page, 20)
    assert_eq(p.offset, 0)
    assert_eq(p.limit, 20)
end)

test("from_query reads ?page=3", function()
    local p = pagination.from_query({ query = { page = "3" } })
    assert_eq(p.page, 3)
    assert_eq(p.offset, 40)  -- (3-1) * 20
end)

test("from_query reads ?per_page=5", function()
    local p = pagination.from_query({ query = { per_page = "5" } })
    assert_eq(p.per_page, 5)
    assert_eq(p.limit, 5)
end)

test("from_query caps per_page at max_per_page", function()
    local p = pagination.from_query({ query = { per_page = "5000" } },
                                     { max_per_page = 50 })
    assert_eq(p.per_page, 50)
end)

test("from_query honors default_per_page", function()
    local p = pagination.from_query({}, { default_per_page = 10 })
    assert_eq(p.per_page, 10)
end)

test("from_query rejects invalid page values, falls back", function()
    local p1 = pagination.from_query({ query = { page = "abc" } })
    assert_eq(p1.page, 1)
    local p2 = pagination.from_query({ query = { page = "-5" } })
    assert_eq(p2.page, 1)
    local p3 = pagination.from_query({ query = { page = "3.5" } })
    assert_eq(p3.page, 1)
end)

test("from_query clamps page to at least 1", function()
    local p = pagination.from_query({ query = { page = "0" } })
    assert_eq(p.page, 1)
end)

-- ── render ──────────────────────────────────────────────────────────

test("render with total=0 has show=false, pages=1", function()
    local r = pagination.render(0, { page = 1, per_page = 20 })
    assert_eq(r.show, false)
    assert_eq(r.pages, 1)
    assert_eq(r.has_prev, false)
    assert_eq(r.has_next, false)
end)

test("render single page (total < per_page) has show=false", function()
    local r = pagination.render(5, { page = 1, per_page = 20 })
    assert_eq(r.pages, 1)
    assert_eq(r.show, false)
end)

test("render computes pages from ceiling", function()
    local r = pagination.render(21, { page = 1, per_page = 10 })
    assert_eq(r.pages, 3)  -- 21/10 rounded up
end)

test("render page=1 of 5: links are [1*,2,3,4,5], no ellipsis", function()
    local r = pagination.render(100, { page = 1, per_page = 20, window = 2 })
    assert_eq(r.pages, 5)
    assert_eq(#r.links, 5)
    assert_eq(r.links[1].active, true)
    assert_eq(r.links[1].page, 1)
    assert_eq(r.links[5].page, 5)
    -- no ellipsis (window=2 covers 1..5 entirely)
    for _, link in ipairs(r.links) do
        assert(not link.ellipsis, "no ellipsis expected")
    end
end)

test("render large total inserts ellipses around current", function()
    local r = pagination.render(400, {
        page = 10, per_page = 20, window = 2
    })
    assert_eq(r.pages, 20)
    -- Expected layout: 1 ... 8 9 10* 11 12 ... 20
    -- That's 9 entries (2 ellipsis + 7 page links)
    assert_eq(#r.links, 9)
    assert_eq(r.links[1].page, 1)
    assert_eq(r.links[2].ellipsis, true)
    -- find the active link
    local active_idx
    for i, link in ipairs(r.links) do
        if link.active then active_idx = i end
    end
    assert(active_idx, "active link present")
    assert_eq(r.links[active_idx].page, 10)
end)

test("render has_prev/has_next reflect boundaries", function()
    local first = pagination.render(100, { page = 1, per_page = 20 })
    assert_eq(first.has_prev, false)
    assert_eq(first.has_next, true)
    local last = pagination.render(100, { page = 5, per_page = 20 })
    assert_eq(last.has_prev, true)
    assert_eq(last.has_next, false)
end)

test("render prev/next URLs include base_url", function()
    local r = pagination.render(100, {
        page = 3, per_page = 20, base_url = "/todos"
    })
    assert_eq(r.prev, "/todos?page=2")
    assert_eq(r.next, "/todos?page=4")
end)

test("render URL preserves existing ? in base_url", function()
    local r = pagination.render(100, {
        page = 3, per_page = 20, base_url = "/search?q=foo"
    })
    assert_eq(r.next, "/search?q=foo&page=4")
end)

test("render URL omits per_page when matching default_per_page", function()
    local r = pagination.render(100, {
        page = 2, per_page = 10, default_per_page = 10, base_url = "/x"
    })
    assert_eq(r.next, "/x?page=3")
end)

test("render URL includes per_page when differing from default", function()
    local r = pagination.render(100, {
        page = 2, per_page = 5, default_per_page = 20, base_url = "/x"
    })
    assert_eq(r.next, "/x?page=3&per_page=5")
end)

test("render clamps page to [1, pages]", function()
    local r = pagination.render(50, { page = 99, per_page = 10 })
    assert_eq(r.page, 5)  -- clamped to max
    local r2 = pagination.render(50, { page = 0, per_page = 10 })
    assert_eq(r2.page, 1)
end)

return { pass = pass, fail = fail }

File: stdlib/js/hull/web/tests/test_pagination.js
// test_pagination.js - Tests for hull:web:pagination

import { pagination } from "hull:web:pagination";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try { fn(); pass++; }
    catch (e) { fail++; console.log(`FAIL: ${name}: ${e.message || e}`); }
}

function assertEq(a, b, msg) {
    if (a !== b) {
        throw new Error(`${msg || ""} expected ${b}, got ${a}`);
    }
}

// ── fromQuery ──────────────────────────────────────────────────────

test("fromQuery defaults to page=1 per_page=20 when no query", () => {
    const p = pagination.fromQuery({});
    assertEq(p.page, 1);
    assertEq(p.per_page, 20);
    assertEq(p.offset, 0);
    assertEq(p.limit, 20);
});

test("fromQuery reads ?page=3", () => {
    const p = pagination.fromQuery({ query: { page: "3" } });
    assertEq(p.page, 3);
    assertEq(p.offset, 40);
});

test("fromQuery reads ?per_page=5", () => {
    const p = pagination.fromQuery({ query: { per_page: "5" } });
    assertEq(p.per_page, 5);
    assertEq(p.limit, 5);
});

test("fromQuery caps per_page at maxPerPage", () => {
    const p = pagination.fromQuery(
        { query: { per_page: "5000" } },
        { maxPerPage: 50 }
    );
    assertEq(p.per_page, 50);
});

test("fromQuery honors defaultPerPage", () => {
    const p = pagination.fromQuery({}, { defaultPerPage: 10 });
    assertEq(p.per_page, 10);
});

test("fromQuery rejects invalid page values, falls back", () => {
    assertEq(pagination.fromQuery({ query: { page: "abc" } }).page, 1);
    assertEq(pagination.fromQuery({ query: { page: "-5" } }).page, 1);
    assertEq(pagination.fromQuery({ query: { page: "3.5" } }).page, 1);
});

test("fromQuery clamps page to at least 1", () => {
    assertEq(pagination.fromQuery({ query: { page: "0" } }).page, 1);
});

// ── render ─────────────────────────────────────────────────────────

test("render with total=0 has show=false, pages=1", () => {
    const r = pagination.render(0, { page: 1, per_page: 20 });
    assertEq(r.show, false);
    assertEq(r.pages, 1);
    assertEq(r.has_prev, false);
    assertEq(r.has_next, false);
});

test("render single page (total < per_page) has show=false", () => {
    const r = pagination.render(5, { page: 1, per_page: 20 });
    assertEq(r.pages, 1);
    assertEq(r.show, false);
});

test("render computes pages from ceiling", () => {
    const r = pagination.render(21, { page: 1, per_page: 10 });
    assertEq(r.pages, 3);
});

test("render page=1 of 5: links are [1*,2,3,4,5], no ellipsis", () => {
    const r = pagination.render(100, { page: 1, per_page: 20, window: 2 });
    assertEq(r.pages, 5);
    assertEq(r.links.length, 5);
    assertEq(r.links[0].active, true);
    assertEq(r.links[0].page, 1);
    assertEq(r.links[4].page, 5);
    for (const link of r.links) {
        if (link.ellipsis) throw new Error("no ellipsis expected");
    }
});

test("render large total inserts ellipses around current", () => {
    const r = pagination.render(400, { page: 10, per_page: 20, window: 2 });
    assertEq(r.pages, 20);
    // Expected: 1 ... 8 9 10* 11 12 ... 20 = 9 entries
    assertEq(r.links.length, 9);
    assertEq(r.links[0].page, 1);
    assertEq(r.links[1].ellipsis, true);
    const activeIdx = r.links.findIndex(l => l.active);
    if (activeIdx < 0) throw new Error("active link present");
    assertEq(r.links[activeIdx].page, 10);
});

test("render has_prev/has_next reflect boundaries", () => {
    const first = pagination.render(100, { page: 1, per_page: 20 });
    assertEq(first.has_prev, false);
    assertEq(first.has_next, true);
    const last = pagination.render(100, { page: 5, per_page: 20 });
    assertEq(last.has_prev, true);
    assertEq(last.has_next, false);
});

test("render prev/next URLs include baseUrl", () => {
    const r = pagination.render(100, {
        page: 3, per_page: 20, baseUrl: "/todos",
    });
    assertEq(r.prev, "/todos?page=2");
    assertEq(r.next, "/todos?page=4");
});

test("render URL preserves existing ? in baseUrl", () => {
    const r = pagination.render(100, {
        page: 3, per_page: 20, baseUrl: "/search?q=foo",
    });
    assertEq(r.next, "/search?q=foo&page=4");
});

test("render URL omits per_page when matching defaultPerPage", () => {
    const r = pagination.render(100, {
        page: 2, per_page: 10, defaultPerPage: 10, baseUrl: "/x",
    });
    assertEq(r.next, "/x?page=3");
});

test("render URL includes per_page when differing from default", () => {
    const r = pagination.render(100, {
        page: 2, per_page: 5, defaultPerPage: 20, baseUrl: "/x",
    });
    assertEq(r.next, "/x?page=3&per_page=5");
});

test("render clamps page to [1, pages]", () => {
    assertEq(pagination.render(50, { page: 99, per_page: 10 }).page, 5);
    assertEq(pagination.render(50, { page: 0, per_page: 10 }).page, 1);
});

export default { pass, fail };
